web/pandas/pdeps/0014-string-dtype.md

```diff
@@ -0,0 +1,388 @@
+subsection below).
+
+This last dtype variant is what users currently (pandas 2.2) get for string data
+when enabling the ``future.infer_string`` option (to enable the behaviour which
+is intended to become the default in pandas 3.0).
+
+## Proposal
+
+To be able to move forward with a string data type in pandas 3.0, this PDEP proposes:
+
+1. For pandas 3.0,  a "string" dtype is enabled by default, which will use PyArrow
+   if installed, and otherwise falls back to an in-house functionally-equivalent
+   (but slower) version.
+2. This default "string" dtype will follow the same behaviour for missing values
+   as other default data types, and use `NaN` as the missing value sentinel.
+3. The version that is not backed by PyArrow can reuse (with minor code
+   additions) the existing numpy object-dtype backed StringArray for its
+   implementation.
+4. Installation guidelines are updated to clearly encourage users to install
+   pyarrow for the default user experience.
+
+Those string dtypes enabled by default will then no longer be considered as
+experimental.
+
+### Default inference of a string dtype
+
+By default, pandas will infer this new string dtype instead of object dtype for
+string data (when creating pandas objects, such as in constructors or IO
+functions).
+
+The existing `future.infer_string` option can be used to opt-in to the future
+default behaviour:
+
+```python
+>>> pd.options.future.infer_string = True
+>>> pd.Series(["a", "b", None])
+0      a
+1      b
+2    NaN
+dtype: string
+```
+
+Right now (pandas 2.2), the existing option only enables the PyArrow-based
+future dtype. For the remaining 2.x releases, this option will be expanded to
+also work when PyArrow is not installed to enable the object-dtype fallback in
+that case.
+
+### Missing value semantics
+
+As mentioned in the background section, the original `StringDtype` has used
+the experimental `pd.NA` sentinel for missing values. In addition to using
+`pd.NA` as the scalar for a missing value, this essentially means
+that:
+
+- String columns follow ["NA-semantics"](https://pandas.pydata.org/docs/user_guide/missing_data.html#na-semantics)
+  for missing values, where `NA` propagates in boolean operations such as
+  comparisons or predicates.
+- Operations on the string column that give a numeric or boolean result use the
+  nullable Integer/Float/Boolean data types (e.g. `ser.str.len()` returns the
+  nullable `'Int64"` / `pd.Int64Dtype()` dtype instead of the numpy `int64`
+  dtype (or `float64` in case of missing values)).
+
+However, up to this date, all other default data types still use `NaN` semantics
+for missing values. Therefore, this proposal says that a new default string
+dtype should also still use the same default missing value semantics and return
+default data types when doing operations on the string column, to be consistent
+with the other default dtypes at this point.
+
+In practice, this means that the default `"string"` dtype will use `NaN` as
```

Comment:
> I'm still -1 on changing this behavior; I do not want to revert "string" back to NumPy nullability semantics

For clarification, do you mean "-1 on using NaN semantics for the default string dtype, regardless of how we name it", or only "-1 on using NaN semantics for the dtype created as `dtype="string"`" ? 

Because it is only the latter that causes the breaking change for anyone already using the nullable string dtype. Assume we would use a different name or different string alias than "string", we could still have a default string dtype (which everyone that was not yet using the nullable `StringDtype` would get by default) that uses the proposed NaN semantics, while not causing a breaking change for the existing users of `dtype="string"` / `dtype=pd.StringDtype()`.

(it's another question whether there is enough support for using a different name, I personally think "string" is the best choice which we should reserve for the default dtype, but first to get a good understanding of your position)

Reply:
> For clarification, do you mean "-1 on using NaN semantics for the default string dtype, regardless of how we name it", or only "-1 on using NaN semantics for the dtype created as `dtype="string"`" ?

Definitely the latter, maybe the former. My expectation with PDEP-10 was that the default pyarrow string would be using pd.NA. If that is too difficult then yea there probably is a compromise on the former, but I do not want to take away the `dtype="string"` functionality from users that has been working all of this time.

Not that it is ideal, but we already have `dtype=str` today and `dtype="string"`; maybe the former becomes the new name for what is being proposed here instead of `string[pyarrow_numpy]` and only change `dtype="string"` to be pyarrow backed without changing nullability semantics?

That doesn't solve the str/"string" discrepancy but I don't think introduces any new problems either

Reply:
>  I do not want to take away the `dtype="string"` functionality from users that has been working all of this time.

We are not "taking away" that functionality, the current revision of the PDEP only asks users to use `dtype=pd.StringDtype(na_value=pd.NA)` instead to continue using the same functionality, and Irv's suggestion would minimize the required code change to use `dtype="String"`

> Not that it is ideal, but we already have `dtype=str` today and `dtype="string"`; maybe the former becomes the new name for what is being proposed here instead of `string[pyarrow_numpy]`

Then what would you propose to show in the `df.dtypes` output? (i.e. the string repr of the dtype) Also "str" instead of "string"? 
That would be an option. In that case, we could also use a separate `StrDtype()` class for those NaN-variants (which also solves the back compat issue for `dtype=pd.StringDtype()`).

Reply:
> Then what would you propose to show in the `df.dtypes` output? (i.e. the string repr of the dtype) Also "str" instead of "string"?

Yea I think this would work. Not saying its an ideal place to be, but I don't think its all that different from what we have today, doesn't require end user changes, and doesn't require a new constructor

Reply:
> In that case, we could also use a separate `StrDtype()` class for those NaN-variants (which also solves the back compat issue for `dtype=pd.StringDtype()`).

Interesting idea, but I think I'm -0.5 on doing that. The proposal here is more in line with PDEP-13's notion of logical types and can segue into that. I think having `pd.StrDtype()` and `pd.StringDtype()` only moves us away from that.

Reply:
> Interesting idea, but I think I'm -0.5 on doing that. The proposal here is more in line with PDEP-13's notion of logical types and can segue into that. I think having `pd.StrDtype()` and `pd.StringDtype()` only moves us away from that.

I'm not sure I agree with you here @WillAyd .  If we had both, and we do logical dtypes, then we pick one or the other.  Similar to how we have to reconcile `dtype=int` versus `dtype=pd.Int64Dtype()` in the future if we use logical types, which would also need to reconcile the `NaN` vs `pd.NA` issue.

Reply:
The idea is that the logical type is decoupled from null handling. If you ignored missing values for a minute, what value do you think there is in having a separate `Str` and `String` type in a type system?

Reply:
> The idea is that the logical type is decoupled from null handling. If you ignored missing values for a minute, what value do you think there is in having a separate `Str` and `String` type in a type system?

No value in a type system.  I just think it may be useful to have this now so we can manage the transition from where we are today to 3.0, and from 3.0 to a future state.

Reply:
This is something we can discuss further in PDEP-13 but the current revision of it proposes consistently having the format of `<TYPE>Dtype(na_marker=pd.NA|"legacy")`, the idea being that long term pd.NA can be used consistently, but we will have a compatability period of "legacy" where you get the mix of np.nan / pd.NaT for NumPy-based types (and still probably pd.NA for any new types).

So `StringDtype(na_marker=np.nan)` is slightly different from that. Maybe asking users to explicitly say `np.nan` instead of `"legacy"` has some downsides from a UI perspective, but my gut feeling is that isn't something we can't solve over time

Reply:
I do want to be wary though of users expecting complete control over the `na_marker` field. I don't see a value-add in trying to support `DatetimeDtype(na_marker=np.nan)` alongside `DatetimeDtype(na_marker=pd.NaT)` nor do I think there would ever be value in `ListDtype(na_marker=np.nan)`

Reply:
> I do want to be wary though of users expecting complete control over the `na_marker` field.

I definitely agree users shouldn't expect complete control over the `na_marker` or `na_value` field. This is not a custom value that can be anything, and we indeed don't want to generalize that to other dtypes in the future. 

Because of that, I have been thinking to not actually allow a user to specify `StringDtype(na_marker=np.nan)` explicitly, but only allow the implicit default of `StringDtype()` (using NaN) or the explicit choice to not have the default with `StringDtype(na_value=pd.NA)`. 
Just to avoid users actually doing `StringDtype(na_marker=np.nan)` (while that is not necessary, when it is possible users will still do it), and thinking this will generalize to other dtypes.

Reply:
I am still overall -1 on changing the default from pd.NA back to np.nan. The latter is not generalizable (not even in our current design) so I don't see how we can consider that the long term solution